lib/curl_setup.h

```diff
@@ -722,4 +722,14 @@ endings either CRLF or LF so 't' is appropriate.
 #define FOPEN_WRITETEXT "w"
 #endif
 
+/* WinSock destroys recv() buffer when send() failed.
+Use workarounds unless explicitly requested to not use.
+Enable workarounds also for Cygwin as Cygwin sockets are wrappers for
+WinSock sockets. https://github.com/curl/curl/issues/657 */
+#if defined(WIN32) || defined(__CYGWIN__)
+#  if !defined(DONT_USE_WINSOCK_WORKAROUNDS)
+#    define USE_WINSOCK_WORKAROUNDS
```

Comment:
The work-around is not really winsock specific. I mean the code works generically for all stacks so it is just an optimization to not do it for everyone. I would to name this define to  `USE_RECV_BEFORE_SEND` or something that tries to say what it does, not which setup that might use it. It is not unthinkable that there are other TCP stacks with the same behavior. I can imagine builds that explicitly switch this on for non-windows builds for testing purposes.